layout: Consider transform for bounding box queries

stevennovaryo opened the pull request:
The recent changes that cached the Scroll Tree present an opportunity to calculate the queries that consider transform and scroll (dubbed as post composite queries) accurately. 

This PR propose a solution for this calculation by noting the lowest scroll tree nodes that would affect a fragment. To do this, each fragment would store a new attribute `spatial_tree_node` -- scroll tree node id that we could use for the query. This referencing is considered because the scroll tree node construction is managed by the fragment itself. Therefore it would ease the managing the possibly stale reference and future query cache invalidation considering the development of incremental layout.

The bounding box query then could transform the bounding content rect of a fragment using the computed current transformation matrix.

Fixes: https://github.com/servo/servo/issues/35768
Testing: Existing and new WPT

stevennovaryo committed:
Handle scroll offset in bounding box query

Signed-off-by: stevennovaryo <steven.novaryo@gmail.com>

stevennovaryo committed:
Calculate CSS transform

Signed-off-by: stevennovaryo <steven.novaryo@gmail.com>

stevennovaryo committed:
Update WPT expectation

Signed-off-by: stevennovaryo <steven.novaryo@gmail.com>

xiaochengh reviewed (approved):
This is really nice! LGTM.

stevennovaryo commented:
Rebased! @xiaochengh

xiaochengh commented on components/layout/fragment_tree/box_fragment.rs:
```diff
@@ -103,6 +104,10 @@ pub(crate) struct BoxFragment {
+
+    /// The lowest scroll tree node that would affect this fragment's border. It is used to
+    /// calculate the post composite queries (e.g., bounding box query).
+    pub scroll_tree_node_id_for_query: AtomicRefCell<Option<ScrollTreeNodeId>>,
```
This adds a new member to every `BoxFragment` and might have a significant impact to the memory consumption. Can we avoid that?

Some of my rough thoughts are:
1. Instead of storing it in a fragment, we can traverse the scroll tree and find the node id during query
2. If that makes the performance too bad, consider caching the result somewhere else

(Thanks for @mrobinson for pointing out!)

stevennovaryo replied:
The issue here is that this information is lost within scroll tree construction, and we does not have a way to identify a specific fragment (except for using a strong pointer). Identifying scroll tree node id from other sources like `DOMLayoutData` might make invalidations a bit complicated. 

I guess we could put this inside `BlockLevelLayoutInfo` which only applies to block level box. My other rough thoughts is using something like `HashTable<Arc<Fragment>, ScrollTreeNodeId>` (with pointer comparison) as stacking context tree info.

xiaochengh replied:
> I guess we could put this inside BlockLevelLayoutInfo which only applies to block level box.

Can we do reduce the stored information even more by storing it for scroll containers only? For other fragments, we can traverse the fragment tree first to find the first scroll container ancestor.

Using a hash map at stacking context tree info looks ugly but otherwise might be fine.

xiaochengh replied:
> For each block level box, it will takes 8 bytes to store it with a pointer like `Option<Box<AtomicRefCell<ScrollTreeNodeId>>>`.

That still sounds too much... Starting with a hash map for scroll containers only sounds better, as we can trade a little bit of tidiness for performance.

Ideally this should be managed by a rare data field, but right not we don't have that on fragments.

Please also start a thread on Zulip to double check with other reviewers.

mrobinson commented on components/layout/query.rs:
```diff
@@ -37,16 +40,111 @@ use style::values::specified::GenericGridTemplateComponent;
+    // TODO(stevennovaryo): Add caching mechanism for this.
+    fn scroll_node_to_root_transform(&self, node_id: &ScrollTreeNodeId) -> LayoutTransform {
+        if let Some(scroll_tree) = self.scroll_tree {
+            let cur_node = scroll_tree.get_node(node_id);
```
Nit:

```suggestion
            let current_node = scroll_tree.get_node(node_id);
```

mrobinson commented on components/layout/query.rs:
```diff
@@ -37,16 +40,111 @@ use style::values::specified::GenericGridTemplateComponent;
+    ) -> Rect<Au> {
+        let layout_rect_union = au_rect_to_f32_rect(rect_to_transform).cast_unit();
+
+        if let Some(transformed_rect) = transform.outer_transformed_rect(&layout_rect_union) {
+            f32_rect_to_au_rect(transformed_rect.to_untyped())
+        } else {
+            rect_to_transform
+        }
+    }
```
```suggestion
        transform.outer_transformed_rect(&layout_rect_union)
            .map(|transformed_rect| 32_rect_to_au_rect(transformed_rect.to_untyped())
            .unwrap_or(rect_to_transform)
    }
```

And in fact, I would simply move these three lines to where they are used twice. I think the function isn't adding too much here.

mrobinson commented on components/layout/query.rs:
```diff
@@ -37,16 +40,111 @@ use style::values::specified::GenericGridTemplateComponent;
+        }
+    }
+
+    fn get_scroll_node_id_for_query(&self, node: ServoLayoutNode<'_>) -> Option<ScrollTreeNodeId> {
```
```suggestion
    fn scroll_node_id_for_query(&self, node: ServoLayoutNode<'_>) -> Option<ScrollTreeNodeId> {
```

mrobinson commented on components/layout/query.rs:
```diff
@@ -37,16 +40,111 @@ use style::values::specified::GenericGridTemplateComponent;
+        PostCompositeQueryHelper { scroll_tree }
+    }
+
+    fn retrieve_box_fragment(fragment: &Fragment) -> Option<&ArcRefCell<BoxFragment>> {
+        match fragment {
+            Fragment::Box(box_fragment) | Fragment::Float(box_fragment) => Some(box_fragment),
+            _ => None,
+        }
+    }
```
I would add this to `Fragment` and call it `fn box_fragment`. It's a bit weird to have the helper here.

mrobinson commented on components/layout/layout_impl.rs:
```diff
@@ -1427,12 +1458,11 @@ impl ReflowPhases {
-                QueryMsg::ClientRectQuery |
                 QueryMsg::ContentBox |
                 QueryMsg::ContentBoxes |
+                QueryMsg::ResolvedStyleQuery |
+                QueryMsg::ScrollingAreaOrOffsetQuery => Self::StackingContextTreeConstruction,
+                QueryMsg::ClientRectQuery |
```
This isn't changing what `ClientRectQuery` requires though. It just removes the curly braces or am I missing something?

stevennovaryo replied:
I believe there is a mixup here. Contrary to its name, `QueryMsg::ClientRectQuery` is actually a query for `ClientTop`, `ClientLeft`, `ClientWidth`, and `ClientHeight`, which ignores the transform. The queries relevant here is `ContentBox` and `ContentBoxes` queries.

mrobinson replied:
No, I mean it seemed like this change is just removing a curly brace, but doesn't seem to be modifying the code otherwise. Perhaps I'm misreading this though...

stevennovaryo replied:
Ah, It is probably due to how does git's shows the changes. Not sure how to show this in the comment, but If we look at the code changes as a whole, it could be seen that it is actually moving `QueryMsg::ContentBox` and `QueryMsg::ContentBoxes` up, adding `StackingContextTreeConstruction` construction requirements.

https://github.com/servo/servo/pull/37871/files#diff-49c9dea0cc9af0161b4681a4c4b1ed81f38f8190bd6854fd2ba6b3d74502b047L1430-R1465

mrobinson replied:
👍

mrobinson commented on components/layout/layout_impl.rs:
```diff
@@ -253,16 +254,26 @@ impl Layout for LayoutThread {
     fn query_content_box(&self, node: TrustedNodeAddress) -> Option<UntypedRect<Au>> {
         let node = unsafe { ServoLayoutNode::new(&node) };
-        process_content_box_request(node)
+        self.process_post_composite_query(node, process_content_box_request)
```
Instead of having the details of the rectangle calculation here, please pass in the StackingContextTree here as an argument to `process_content_box_request`.

mrobinson commented on components/layout/layout_impl.rs:
```diff
@@ -1155,6 +1166,26 @@ impl LayoutThread {
             .force_stylesheet_origins_dirty(sheet_origins_affected_by_device_change);
     }
+
+    /// Main entry point for queries that require [PostCompositeQueryHelper].
+    fn process_post_composite_query<T>(
+        &self,
+        node: ServoLayoutNode<'_>,
+        process_query_internals: fn(
+            node: ServoLayoutNode<'_>,
+            helper: PostCompositeQueryHelper,
+        ) -> T,
+    ) -> T {
+        let stacking_context_tree = self.stacking_context_tree.borrow();
+
+        let helper = PostCompositeQueryHelper::new(
+            stacking_context_tree
+                .as_ref()
+                .map(|tree| &tree.compositor_info.scroll_tree),
+        );
+
+        process_query_internals(node, helper)
+    }
```
Please remove this function as it is unnecessary. Instead let's create the `PostCompositeQueryHelper` in `query.rs`.

stevennovaryo replied:
Sure, it does seems overcomplicating the flow, thanks!

mrobinson commented on components/layout/layout_impl.rs:
```diff
@@ -253,16 +254,26 @@ impl Layout for LayoutThread {
     fn query_content_boxes(&self, node: TrustedNodeAddress) -> Vec<UntypedRect<Au>> {
         let node = unsafe { ServoLayoutNode::new(&node) };
-        process_content_boxes_request(node)
+        self.process_post_composite_query(node, process_content_boxes_request)
```
Instead of having the details of the rectangle calculation here, please pass in the StackingContextTree here as an argument to process_content_boxes_request.

stevennovaryo replied:
Got it, thanks!

mrobinson commented on components/layout/query.rs:
```diff
@@ -37,16 +40,111 @@ use style::values::specified::GenericGridTemplateComponent;
+        Some(self.scroll_node_to_root_transform(&scroll_tree_node_id))
+    }
+
+    /// Traverse a scroll node to its root to calculate the transform.
+    // TODO(stevennovaryo): Add caching mechanism for this.
+    fn scroll_node_to_root_transform(&self, node_id: &ScrollTreeNodeId) -> LayoutTransform {
```
Let's just make this a free function called `fn cumulative_scroll_node_transform_for_node(node: ServoLayoutNode<'_>)` and combine it with `get_scroll_node_id_for_query`. The name `PostCompositeQueryHelper` is really general and this is really just doing one thing (getting the cumulative scroll node transformation for a node).

stevennovaryo replied:
Sorry, the scope and state of these functions does seems quite inappropriate. I have changed `cumulative_scroll_node_transform_for_node` to be a free function of `PostCompositeQueryHelper`. That feels appropriate since it is still a subset of post composite queries. I have also combined `get_scroll_node_id_for_query` with `layout_node_to_root_transform`, seems that is the most appropriate approach in the current state.

mrobinson commented on components/layout/query.rs:
```diff
@@ -55,18 +153,38 @@ pub fn process_content_box_request(node: ServoLayoutNode<'_>) -> Option<Rect<Au>
-        .collect()
+        .map(|rect| rect.to_untyped());
+
+    let maybe_transform = helper.layout_node_to_root_transform(node);
```
Here and above which will make code flow a lot simpler.


```suggestion
    let Some(transform) = helper.layout_node_to_root_transform(node) else {
        return content_boxes.collect()
    }
```

mrobinson commented on components/layout/display_list/stacking_context.rs:
```diff
@@ -992,8 +994,11 @@ impl BoxFragment {
+
         let new_spatial_id = stacking_context_tree.push_reference_frame(
             reference_frame_data.origin.to_webrender(),
+            frame_origin_for_query,
```
Is this truly the right value to use for this? It isn't taking into account any of the information from ancestors such as scroll offsets and transforms...

stevennovaryo replied:
Yes, we need this find the correct offset (relative to the document) when a transformation happens. And, the offset of the given rect in the queries was without any transform (and in fact, is using the cumulative box as well).

mrobinson reviewed (changes requested):
I could see attaching the scroll tree node to the `BoxFragment` being good for memory usage if it ultimately allows us to remove the expensive hit testing data structure we sent to the compositor.

stevennovaryo commented on components/layout/layout_impl.rs:
```diff
@@ -253,16 +254,26 @@ impl Layout for LayoutThread {
+    /// that returns union of it.
+    ///
+    /// Part of <https://drafts.csswg.org/cssom-view-1/#element-get-the-bounding-box>
+    /// TODO(stevennovaryo): Rename and parameterize the function, allowing padding area
+    ///                      query and possibly, query without consideration of transform.
```
It is useful for intersection observer, specifically in getting the root intersection rectangle https://w3c.github.io/IntersectionObserver/#intersectionobserver-root-intersection-rectangle. Where we need to use padding rect instead of border rect in the query. 

As for query without transformation, I suppose it is just a preliminary thoughts. Currently, I think that it is better to kept them separate.

stevennovaryo committed:
Apply suggestion

Signed-off-by: stevennovaryo <steven.novaryo@gmail.com>

mrobinson commented on components/layout/query.rs:
```diff
@@ -37,16 +40,112 @@ use style::values::specified::GenericGridTemplateComponent;
+    scroll_tree: Option<&'dom ScrollTree>,
+}
+
+impl<'dom> PostCompositeQueryHelper<'dom> {
```
After a bit of discussion, I think we should remove the `PostCompositeQueryHelper` struct entirely. There is no "composite" step in Servo, so this name is pretty confusing. Instead I think `layout_node_to_root_transform` should probably just do everything here taking the `ScrollTree` as an argument. Sorry for the delay and confusion on getting back to you to you on this.

stevennovaryo committed:
Remove PostCompositeQuery

Signed-off-by: Jo Steven Novaryo <jo.steven.novaryo@huawei.com>

mrobinson committed:
Some minor changes:

- Rename and remove some functions to avoid "post_composite" naming.
  Servo doesn't have a composite step like other web engines.
- Move cumulative transform function to `ScrolTree` itself.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson committed:
Also assign `scroll_node_id` to float fragments

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

stevennovaryo committed:
Fix spatial node assigning

Signed-off-by: Jo Steven Novaryo <jo.steven.novaryo@huawei.com>

stevennovaryo committed:
Add minimal WPT for getBoundingBox with scroll containers

Signed-off-by: Jo Steven Novaryo <jo.steven.novaryo@huawei.com>

stevennovaryo commented:
> The assertion seems to have caught an issue with float fragments which weren't having scroll nodes assigned properly.

Thanks for catching the issue with float fragments, it should be assigned with correct nodes as well.

> I suspect there is going to be an issue with `display: none` content as well.

Added minimal WPTs for scroll, `display: none` and `display: content`.

It should be correct as long as a BoxFragment went through the stacking context construction phases. If not, it could find the closest parent that does and calculate the transformation from there. Since fragment that does not went through that phases is not creating any frame, it would present a correct calculation.
